app/lib/frontend/backend.dart

```diff
@@ -159,12 +160,66 @@ class Backend {
     assert(repository.supportsDownloadUrl);
     return repository.downloadUrl(package, version);
   }
+
+  /// Stores a [models.UrlNonce] entry in the Datastore and returns its id that
+  /// can be used in client-communication, e.g. sending via e-mail.
+  ///
+  /// Duplicate detection relies on the serialized JSON form of [parameters],
+  /// and it is expected that callers will order the keys consistently.
+  ///
+  /// When multiple duplicate is detected, the return value becomes `null`.
+  Future<String> createUrlNonce(String action, Map<String, dynamic> parameters,
+      {Duration expires = const Duration(days: 1)}) async {
+    final id = _uuid.v4().toString();
+    final now = new DateTime.now().toUtc();
+    final verification = new models.UrlNonce()
+      ..parentKey = db.emptyKey
+      ..id = id
+      ..created = now
+      ..expires = now.add(expires)
+      ..action = action
+      ..parameters = parameters
+      ..updateHash();
+
+    final query = db.query<models.UrlNonce>()
+      ..filter('dedupHash =', verification.dedupHash);
+    // TODO: match parameters too for 100% certainty
+    // The chance of mis-classification is low, but we should eventually address it.
+    final count = await query
+        .run()
+        .where((n) => n.isActive() && n.action == action)
+        .length;
+    if (count > 5) {
+      return null;
+    }
+
+    await db.commit(inserts: [verification]);
+    return id;
+  }
+
+  /// Updates the [models.UrlNonce] entry (identified by [action] and [id]), and
+  /// return it parameters (or null if it has been expired).
+  Future<Map<String, dynamic>> confirmUrlNonce(String action, String id) async {
+    return await db.withTransaction((tx) async {
+      final models.UrlNonce v =
+          (await tx.lookup([db.emptyKey.append(models.UrlNonce, id: id)]))
+              .single;
+      final now = new DateTime.now().toUtc();
+      if (v == null || !v.isActive() || v.action != action) {
```

Comment:
If it's been confirmed within the past 10 min we should allow it through again...